support/cas-server-support-x509-core/src/main/java/org/apereo/cas/adaptors/x509/authentication/principal/AbstractX509PrincipalResolver.java

```diff
@@ -131,8 +135,47 @@ protected String getAlternatePrincipal(final X509Certificate certificate) {
             if (subjectPrincipal != null) {
                 attributes.put("subjectX500Principal", CollectionUtils.wrapList(subjectPrincipal.getName()));
             }
+            try {
+                val rfc822Email = getRFC822EmailAddress(certificate.getSubjectAlternativeNames());
+                if (rfc822Email != null) {
+                    attributes.put("x509Rfc822Email", CollectionUtils.wrapList(rfc822Email));
+                }
+            } catch (final CertificateParsingException e) {
+                LOGGER.warn("Error parsing subject alternative names to get rfc822 email {}", e.getMessage());
+            }
         }
         return attributes;
     }
 
+    /**
+     * Get Email Address.
+     *
+     * @param subjectAltNames list of subject alternative name values encoded as collection of Lists with two elements in each List containing type and value.
+     * @return String email address or null if the item passed in is not type 1 (rfc822Name)
+     * as expected to be returned by implementation of {@code X509Certificate.html#getSubjectAlternativeNames}
+     * @see <a href="http://docs.oracle.com/javase/7/docs/api/java/security/cert/X509Certificate.html#getSubjectAlternativeNames()">
+     * X509Certificate#getSubjectAlternativeNames</a>
+     */
+    protected String getRFC822EmailAddress(final Collection<List<?>> subjectAltNames) {
+        try {
+            if (subjectAltNames == null) {
+                return null;
+            }
+            Optional<List<?>> email = subjectAltNames.stream().filter(s -> ((Integer) s.get(0)) == RFC822_SAN_TYPE).findFirst();
+            if (email.isPresent()) {
+                return (String) email.get().get(1);
+            }
+        } catch (final ArrayIndexOutOfBoundsException e) {
```

Comment:
This is very odd; not the sort of thing you should catch. Can this be prevented?

Reply:
I was just trying to use a stream operation while still getting the behavior of the UPN resolver where the code does a check to make sure there are two items in the list. I don't think it is likely to happen and it may be an unnecessary check. I could either remove it or change it to a loop with an if check.

Reply:
Yes please remove and revisit. You could check for length before accessing elements at assumed positions, and if you find any other spot that does the same, please update those as well.